File: services/visualization/src/layer/domain/spec/Transform.ts
import { CoordinatesPositionEncoding } from '../../../encoding/domain/spec/PositionEncoding.js'
import { SizeEncoding } from '../../../encoding/domain/spec/SizeEncoding.js'

export enum TransformType {
  cluster = 'cluster',
  roomCluster = 'roomCluster',
  unit = 'unit',
  neighborCount = 'neighborCount',
  place = 'place',
  placeAround = 'placeAround',
  groupBy = 'groupBy'
}

interface BaseTransform<TType extends TransformType> {
  type: TType
}

interface ClusterTransform extends BaseTransform<TransformType.cluster> {
  radius: number
}

interface NeighborCountTransform extends BaseTransform<TransformType.neighborCount> {
  radius: number
}

interface GroupByTransform extends BaseTransform<TransformType.groupBy> {
  fields: string[]
}

interface RoomClusterTransform extends BaseTransform<TransformType.roomCluster> {
  cellSize?: number
  roomCoordinates?: CoordinatesPositionEncoding
  itemGroup?: {
    field: string
  },
  floor?: {
    field: string
  }
}

enum SortOder {
  ASC = 'ASC',
  DESC = 'DESC'
}

interface UnitTransform extends BaseTransform<TransformType.unit> {
  group: {
    field: string
    limit?: number
  }
  sort?: {
    field: string
    order?: SortOder
  }
}

interface PlaceTransform extends BaseTransform<TransformType.place> {
  padding?: number
  zoom?: number
}

interface PlaceAroundTransform extends BaseTransform<TransformType.placeAround> {
  parentId: {
    field: string
  },
  spacing?: SizeEncoding
  row?: {
    minSize: SizeEncoding
  }
}

export type Transform = ClusterTransform | RoomClusterTransform | UnitTransform | NeighborCountTransform 
                        | PlaceTransform | GroupByTransform | PlaceAroundTransform

File: services/dataset/src/main/java/hopara/dataset/transform/Transform.java
package hopara.dataset.transform;

import java.io.Serializable;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

import hopara.dataset.column.Columns;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes({
  @JsonSubTypes.Type(value=ClusterTransform.class, name = "CLUSTER"),
  @JsonSubTypes.Type(value=RoomClusterTransform.class, name = "ROOM_CLUSTER"),
  @JsonSubTypes.Type(value=NeighborCountTransform.class, name = "NEIGHBOR_COUNT"),
  @JsonSubTypes.Type(value=UnitTransform.class, name = "UNIT"),
  @JsonSubTypes.Type(value=PlaceTransform.class, name = "PLACE"),
  @JsonSubTypes.Type(value=PlaceAroundTransform.class, name = "PLACE_AROUND"),
  @JsonSubTypes.Type(value=GroupByTransform.class, name = "GROUP_BY")
})
@JsonIgnoreProperties(ignoreUnknown = true)
public abstract class Transform implements Serializable  {
    public abstract TransformType getType();
    public abstract Columns getColumns(Columns viewColumns);
}
